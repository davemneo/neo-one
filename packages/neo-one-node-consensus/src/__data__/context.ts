import { common } from '@neo-one/client-common';
import _ from 'lodash';
import { cloneBlockSent, cloneInitial, Context } from '../context';
import { keys } from './keys';

const validators = keys.map(({ publicKey }) => publicKey);
const previousHash = common.stringToUInt256('0xd42561e3d30e15be6400b6df2f328e02d2bf6354c41dce433bc57687c82144bf');

const expectedView = _.range(0, 7).map(() => 0);
const blockReceivedTimeSeconds = 1513018010;

const viewNumber = 0;
const primaryIndex = 1;
const blockIndex = primaryIndex;
const backupIndex = 2;
const backupContext = new Context({
  type: 'backup',
  previousHash,
  blockIndex,
  viewNumber,
  myIndex: backupIndex,
  primaryIndex,
  expectedView,
  validators,
  blockReceivedTimeSeconds,
});

const backupPrivateKey = keys[backupIndex].privateKey;

const primaryContext = new Context({
  type: 'primary',
  previousHash,
  blockIndex,
  viewNumber,
  myIndex: primaryIndex,
  primaryIndex,
  expectedView,
  validators,
  blockReceivedTimeSeconds,
});

const primaryPrivateKey = keys[primaryIndex].privateKey;

const blockSentBackupContext = cloneBlockSent(backupContext);
const initialBackupContext = cloneInitial(backupContext, {
  type: 'backup',
  viewNumber,
  primaryIndex,
});

const blockSentPrimaryContext = cloneBlockSent(primaryContext);
const initialPrimaryContext = cloneInitial(primaryContext, {
  type: 'primary',
  viewNumber,
  primaryIndex,
});

export const context = {
  backupPrivateKey,
  primaryPrivateKey,
  previousHash,
  blockSentBackupContext,
  initialBackupContext,
  blockSentPrimaryContext,
  initialPrimaryContext,
};
