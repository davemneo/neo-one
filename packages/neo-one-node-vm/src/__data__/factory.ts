import { common, ECPoint, InputModelAdd, OutputModelAdd, UInt160, UInt256 } from '@neo-one/client-common';
import {
  Account,
  AccountAdd,
  Asset,
  AssetAdd,
  Block,
  BlockBaseAdd,
  BufferAttribute,
  BufferAttributeAdd,
  ChangeViewConsensusMessage,
  ConsensusPayload,
  ConsensusPayloadAdd,
  Contract,
  ContractAdd,
  ECPointAttribute,
  ECPointAttributeAdd,
  Header,
  HeaderAdd,
  Input,
  Output,
  UInt160Attribute,
  UInt160AttributeAdd,
  UInt256Attribute,
  UInt256AttributeAdd,
  Validator,
  ValidatorAdd,
  Witness,
  WitnessAdd,
} from '@neo-one/node-core';
import BN from 'bn.js';

const internalData: ReadonlyArray<{
  readonly version: number;
  readonly hash256: UInt256;
  readonly hash160: UInt160;
  readonly asset: UInt256;
  readonly address: UInt160;
  readonly script: Buffer;
  readonly merkleRoot: UInt256;
  readonly timestamp: number;
  readonly consensusData: BN;
  readonly nextConsensus: UInt160;
  readonly verification: Buffer;
  readonly invocation: Buffer;
  readonly publicKey: ECPoint;
}> = [
  {
    version: 1,
    hash160: common.asUInt160(Buffer.from('f3812db982f3b0089a21a278988efeec6a027b25', 'hex')),
    hash256: common.hexToUInt256('0x07ae9ba488514e014fdbf166108b038fe94c54dbaedebfb6e676504b2a32cb47'),
    asset: common.hexToUInt256('0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234'),
    address: common.asUInt160(Buffer.from('197ff6783d512a740d42f4cc4f5572955fa44c95', 'hex')),
    script: Buffer.from([0x01]),
    merkleRoot: common.hexToUInt256('0x823e4d905b88d042e5a4038668a0068e331b673d79ebc91e3c9ecc5828a28de4'),
    consensusData: new BN(1),
    timestamp: 0,
    nextConsensus: common.asUInt160(Buffer.from('197ff6783d512a740d42f4cc4f5572955fa44c95', 'hex')),
    verification: Buffer.from('asdf08asdfljawerlk', 'hex'),
    invocation: Buffer.from('savpoudfglkwerl;wqer', 'hex'),
    publicKey: common.bufferToECPoint(
      Buffer.from('02028a99826edc0c97d18e22b6932373d908d323aa7f92656a77ec26e8861699ef', 'hex'),
    ),
  },
  {
    version: 1,
    hash160: common.asUInt160(Buffer.from('197ff6783d512a740d42f4cc4f5572955fa44c95', 'hex')),
    hash256: common.hexToUInt256('0x9bf50b229d070a243ec43f9979e1922d8275cbe2676b811e614ef4c3271ab7a9'),
    asset: common.hexToUInt256('0x0621113df436c180a47b833a2814008dc4f332915b22253651593a40a342e0fb'),
    address: common.asUInt160(Buffer.from('f3812db982f3b0089a21a278988efeec6a027b25', 'hex')),
    script: Buffer.from([0x02]),
    merkleRoot: common.hexToUInt256('0x7e2a450ffecb98d95380185a7f099697e0679a501403ae453cd79d171b40cc5f'),
    consensusData: new BN(2),
    timestamp: 1,
    nextConsensus: common.asUInt160(Buffer.from('f3812db982f3b0089a21a278988efeec6a027b25', 'hex')),
    verification: Buffer.from('sfgjuioyrsfgvacsf', 'hex'),
    invocation: Buffer.from('4356wr6i8fdgdnbsd', 'hex'),
    publicKey: common.bufferToECPoint(
      Buffer.from('031d8e1630ce640966967bc6d95223d21f44304133003140c3b52004dc981349c9', 'hex'),
    ),
  },
  {
    version: 1,
    hash160: common.asUInt160(Buffer.from('197ff6783d512a740d42f4cc4f5572955fa44c95', 'hex')),
    hash256: common.hexToUInt256('0x963edfc071a8708711cc3ed842dc4a06b5357d7330b3df7d147e7c150580f29c'),
    asset: common.hexToUInt256('0x0621113df436c180a47b833a2814008dc4f332915b22253651593a40a342e0fb'),
    address: common.asUInt160(Buffer.from('f3812db982f3b0089a21a278988efeec6a027b25', 'hex')),
    script: Buffer.from([0x02]),
    merkleRoot: common.hexToUInt256('0x7e2a450ffecb98d95380185a7f099697e0679a501403ae453cd79d171b40cc5f'),
    consensusData: new BN(2),
    timestamp: 1,
    nextConsensus: common.asUInt160(Buffer.from('197ff6783d512a740d42f4cc4f5572955fa44c95', 'hex')),
    verification: Buffer.from('fgntyotfd', 'hex'),
    invocation: Buffer.from('ert7rsxdgfs', 'hex'),
    publicKey: common.bufferToECPoint(
      Buffer.from('02232ce8d2e2063dce0451131851d47421bfc4fc1da4db116fca5302c0756462fa', 'hex'),
    ),
  },
];

// tslint:disable:no-let
let index = 1;
const previousData = internalData[index - 1];
const currentData = internalData[index];
// tslint:enable:no-let

const incrementData = () => {
  if (index === internalData.length - 1) {
    index = 1;
  }
  index = index + 1;
};

const resetDataIndex = () => {
  index = 1;
};

const createAccount = (options: Partial<AccountAdd> = {}) =>
  new Account({
    version: currentData.version,
    hash: currentData.hash160,
    ...options,
  });

const createAsset = (options: Partial<AssetAdd> = {}) =>
  new Asset({
    hash: currentData.hash256,
    type: 0x60,
    name: 'test',
    amount: new BN(1),
    precision: 8,
    owner: createECPoint(0x20),
    admin: previousData.hash160,
    issuer: currentData.hash160,
    expiration: 20,
    ...options,
  });

const createBufferAttribute = (options: Partial<BufferAttributeAdd> = {}) =>
  new BufferAttribute({
    usage: 0x81,
    value: currentData.asset,
    ...options,
  });

const createBlock = (options: Partial<BlockBaseAdd> = {}) =>
  new Block({
    previousHash: previousData.hash256,
    merkleRoot: currentData.merkleRoot,
    timestamp: 0,
    index: 0,
    consensusData: new BN(3),
    nextConsensus: currentData.nextConsensus,
    transactions: [],
    script: createWitness(),
    hash: currentData.hash256,
    ...options,
  });

const createECPointAttribute = (options: Partial<ECPointAttributeAdd> = {}) =>
  new ECPointAttribute({
    usage: 0x02,
    value: createECPoint(),
    ...options,
  });

const createUInt160Attribute = (options: Partial<UInt160AttributeAdd> = {}) =>
  new UInt160Attribute({
    usage: 0x20,
    value: currentData.hash160,
    ...options,
  });

const createUInt256Attribute = (options: Partial<UInt256AttributeAdd> = {}) =>
  new UInt256Attribute({
    usage: 0x00,
    value: currentData.hash256,
    ...options,
  });

const createOutput = (options: Partial<OutputModelAdd> = {}) =>
  new Output({
    asset: currentData.asset,
    value: new BN(0),
    address: currentData.address,
    ...options,
  });

const createConsensusPayload = (options: Partial<ConsensusPayloadAdd> = {}) =>
  new ConsensusPayload({
    version: currentData.version,
    previousHash: currentData.hash256,
    blockIndex: index,
    validatorIndex: index + 10,
    consensusMessage: new ChangeViewConsensusMessage({
      newViewNumber: 5,
      viewNumber: 3,
    }),
    script: createWitness(),
    ...options,
  });

const createContract = (options: Partial<ContractAdd> = {}) =>
  new Contract({
    version: currentData.version,
    script: currentData.script,
    parameterList: [],
    returnType: 0x01,
    contractProperties: 0x01,
    name: 'test',
    codeVersion: '1.0',
    author: 'Herby Hind',
    email: 'HerbyHind@neotracker.io',
    description: 'test contract',
    ...options,
  });
const createECPoint = (byte?: number) =>
  common.asECPoint(Buffer.from([...Array(33)].map(() => (byte === undefined ? 0x01 : byte))));

const createInput = (options: Partial<InputModelAdd> = {}) =>
  new Input({
    hash: currentData.asset,
    index,
    ...options,
  });

const createHeader = (options: Partial<HeaderAdd> = {}): Header =>
  new Header({
    previousHash: previousData.hash256,
    merkleRoot: currentData.merkleRoot,
    timestamp: currentData.timestamp,
    index,
    consensusData: currentData.consensusData,
    nextConsensus: currentData.nextConsensus,
    script: createWitness(),
    hash: currentData.hash256,
    ...options,
  });

const createUInt160 = () => currentData.hash160;
const createUInt256 = () => currentData.hash256;

const createValidator = (options: Partial<ValidatorAdd> = {}) =>
  new Validator({
    publicKey: currentData.publicKey,
    ...options,
  });

const createWitness = (options: Partial<WitnessAdd> = {}): Witness =>
  new Witness({
    verification: currentData.verification,
    invocation: currentData.invocation,
    ...options,
  });

export const factory = {
  createAccount,
  createAsset,
  createBlock,
  createBufferAttribute,
  createConsensusPayload,
  createContract,
  createECPoint,
  createECPointAttribute,
  createHeader,
  createInput,
  createOutput,
  createUInt160,
  createUInt160Attribute,
  createUInt256,
  createUInt256Attribute,
  createValidator,
  createWitness,
  incrementData,
  resetDataIndex,
};
