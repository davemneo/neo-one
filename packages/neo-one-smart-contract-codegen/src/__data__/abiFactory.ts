import { ABI, ABIDefault, ABIReturn } from '@neo-one/client-common';

// tslint:disable-next-line:no-any
const getCombinations = (list: ReadonlyArray<ReadonlyArray<any>>) => {
  const floorsIn = list.reduceRight<ReadonlyArray<number>>((acc, right) => [right.length * acc[0]].concat(acc), [1]);
  const max = floorsIn[0];
  const floors = floorsIn.slice(1);

  const iteratorFuncs = list.map((item, index) =>
    index === list.length - 1
      ? (value: number) => item[value % item.length]
      : (value: number) => item[Math.floor(value / floors[index]) % item.length],
  );

  return [...Array(max)].map((_, index) => iteratorFuncs.map((func) => func(index)));
};

const abiFunctionOptions: ReadonlyArray<{ readonly [key in string]?: boolean }> = [
  { constant: true },
  { send: true },
  { sendUnsafe: true },
  { sendUnsafe: true, receive: true },
  { receive: true },
  { claim: true },
  { refundAssets: true },
  { completeSend: true },
  {},
];

const abiParameterOptions: ReadonlyArray<{ readonly [key in string]?: boolean | ABIDefault }> = [
  { default: { type: 'sender' } },
  { rest: true },
  { default: { type: 'sender' }, rest: true },
  {},
];

const abiReturns: ReadonlyArray<ABIReturn> = [
  {
    type: 'ForwardValue',
  },
  {
    type: 'Boolean',
    optional: true,
  },
  {
    type: 'Boolean',
    forwardedValue: true,
  },
  {
    type: 'Integer',
    decimals: 4,
  },
  {
    type: 'Array',
    value: {
      type: 'Boolean',
    },
  },
  {
    type: 'Map',
    key: { type: 'Integer', decimals: 0 },
    value: { type: 'Boolean' },
  },
  {
    type: 'Object',
    properties: {
      property: { type: 'Boolean' },
    },
  },
];

const boolReturnType: ABIReturn = { type: 'Boolean' };

const createBaseABIFunctions = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiFunctionOptions]);

  return combinations.map(([abiReturn, abiFuncProps]) => {
    const funcPropNames = Object.keys(abiFuncProps);
    const name = `${abiReturn.type}${`${funcPropNames.length === 0 ? '' : '_'}${funcPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          returnType: abiReturn,
          ...abiFuncProps,
        },
      ],
    };
  });
};

const createBaseABIEvents = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name: 'foo',
          returnType: boolReturnType,
        },
      ],
      events: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
        },
      ],
    };
  });
};

const createBaseABIFunctionParameters = (): ReadonlyArray<ABI> => {
  const combinations = getCombinations([abiReturns, abiParameterOptions]);

  return combinations.map(([abiReturn, abiParamProps]) => {
    const paramPropNames = Object.keys(abiParamProps);
    const name = `${abiReturn.type}${`${paramPropNames.length === 0 ? '' : '_'}${paramPropNames.join('_')}`}`;

    return {
      functions: [
        {
          name,
          parameters: [
            {
              name: 'arg',
              ...abiReturn,
              ...abiParamProps,
            },
          ],
          returnType: boolReturnType,
        },
      ],
    };
  });
};

export const abiFactory = {
  createBaseABIFunctions,
  createBaseABIFunctionParameters,
  createBaseABIEvents,
};
