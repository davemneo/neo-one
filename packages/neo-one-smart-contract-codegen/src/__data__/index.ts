export * from './nep5';
export * from './testUtils';
export * from './abiFactory';
