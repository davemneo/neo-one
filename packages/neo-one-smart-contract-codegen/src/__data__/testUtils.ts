import { ABI } from '@neo-one/client-common';
import { genSmartContractTypes } from '../types';

const testABI = (abi: ABI, name: string) => {
  test(name, () => {
    expect({
      inputABI: abi,
      types: genSmartContractTypes(name, abi),
    }).toMatchSnapshot();
  });
};

export const testUtils = {
  testABI,
};
