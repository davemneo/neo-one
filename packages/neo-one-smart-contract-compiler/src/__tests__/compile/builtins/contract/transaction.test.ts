import { Attribute, AttributeUsageModel as AttributeUsage, common, Input, Output } from '@neo-one/client-common';
import BigNumber from 'bignumber.js';
import { helpers, keys } from '../../../../__data__';
import { DiagnosticCode } from '../../../../DiagnosticCode';

describe('Transaction', () => {
  test('properties', async () => {
    const node = await helpers.startNode();
    const block = await node.readClient.getBlock(0);
    const { transaction: invocationTransaction } = await node.executeString(
      `
      import { Hash256, TransactionType, Transaction, } from '@neo-one/smart-contract';

      const transaction = Transaction.for(Hash256.from('${block.transactions[0].hash}'));

      assertEqual(transaction.type, TransactionType.Miner);
    `,
    );
    const { transaction, confirmed } = await node.client.transfer(
      [
        {
          to: keys[0].address,
          amount: new BigNumber(10),
          asset: common.NEO_ASSET_HASH,
        },
      ],
      {
        attributes: [
          {
            usage: 'Description',
            data: Buffer.from('Hello World', 'utf8').toString('hex'),
          },
        ],
      },
    );
    await confirmed();

    const getUsage = (attribute: Attribute) => AttributeUsage[attribute.usage];

    const checkAttribute = (idx: number, attribute: Attribute) => `
      attribute = attributes[${idx}];

      assertEqual(attribute.usage, ${getUsage(attribute)});
      assertEqual(attribute.data.equals(${helpers.getBufferHash(attribute.data)}), true)
      assertEqual(attribute instanceof AttributeBase, true);
    `;

    const checkInput = (idx: number, input: Input) => `
      input = inputs[${idx}];

      assertEqual(input.hash.equals(Hash256.from('${input.hash}')), true)
      assertEqual(input.index, ${input.index})
      assertEqual(input instanceof Input, true);
    `;

    const checkOutput = (idx: number, output: Output) => `
      output = outputs[${idx}];

      assertEqual(output.asset.equals(Hash256.from('${output.asset}')), true);
      assertEqual(output.value, ${helpers.getDecimal(output.value)});
      assertEqual(output.address.equals(Address.from('${output.address}')), true);
      assertEqual(output instanceof Output, true);
    `;

    const references = await Promise.all(transaction.inputs.map(async (input) => node.readClient.getOutput(input)));

    await node.executeString(`
      import { TransactionBase, TransactionType, Transaction, Address, Hash256, Attribute, Output, Input, InvocationTransaction, AttributeBase } from '@neo-one/smart-contract';

      const invocationTransaction = Transaction.for(Hash256.from('${
        invocationTransaction.hash
      }')) as InvocationTransaction;
      assertEqual(invocationTransaction.script.equals(${helpers.getBufferHash(invocationTransaction.script)}), true);

      const transaction = Transaction.for(Hash256.from('${transaction.hash}')) as InvocationTransaction;

      assertEqual(transaction.type, TransactionType.Invocation);

      const attributes = transaction.attributes;
      assertEqual(attributes.length, ${transaction.attributes.length});

      let attribute: Attribute;
      ${transaction.attributes.map((attribute, idx) => checkAttribute(idx, attribute)).join('')}

      const inputs = transaction.inputs;
      assertEqual(inputs.length, ${transaction.inputs.length});

      let input: Input;
      ${transaction.inputs.map((input, idx) => checkInput(idx, input)).join('')}

      let outputs = transaction.outputs;
      assertEqual(outputs.length, ${transaction.outputs.length});

      let output: Output;
      ${transaction.outputs.map((output, idx) => checkOutput(idx, output)).join('')}

      outputs = transaction.unspentOutputs;
      assertEqual(outputs.length, ${transaction.outputs.length});

      ${transaction.outputs.map((output, idx) => checkOutput(idx, output)).join('')}

      outputs = transaction.references;
      assertEqual(outputs.length, ${references.length});

      ${references.map((output, idx) => checkOutput(idx, output)).join('')}

      assertEqual(transaction instanceof TransactionBase, true);
    `);
  });

  test('cannot be implemented', async () => {
    helpers.compileString(
      `
      import { TransactionBase } from '@neo-one/smart-contract';

      class MyTransaction implements TransactionBase {
      }
    `,
      { type: 'error' },
    );
  });

  test('invalid reference', () => {
    helpers.compileString(
      `
      import { Transaction } from '@neo-one/smart-contract';

      const for = Transaction.for;
    `,
      { type: 'error', code: DiagnosticCode.InvalidBuiltinReference },
    );
  });

  test('invalid "reference"', () => {
    helpers.compileString(
      `
      import { Transaction } from '@neo-one/smart-contract';

      const for = Transaction['for'];
    `,
      { type: 'error', code: DiagnosticCode.InvalidBuiltinReference },
    );
  });

  test('invalid reference - object', () => {
    helpers.compileString(
      `
      import { Transaction } from '@neo-one/smart-contract';

      const { for } = Transaction;
    `,
      { type: 'error', code: DiagnosticCode.InvalidBuiltinReference },
    );
  });
});
