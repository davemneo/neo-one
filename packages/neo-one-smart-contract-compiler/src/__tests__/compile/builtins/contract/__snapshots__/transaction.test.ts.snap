// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction cannot be implemented 1`] = `
"snippetCode.ts (4,13): Class 'MyTransaction' incorrectly implements interface 'TransactionBase'.
  Property 'hash' is missing in type 'MyTransaction'.

      2 |       import { TransactionBase } from '@neo-one/smart-contract';
      3 | 
    > 4 |       class MyTransaction implements TransactionBase {
        |             ^
      5 |       }
      6 |     
"
`;

exports[`Transaction invalid "reference" 1`] = `
"snippetCode.ts (4,31): Builtin properties cannot be referenced

      2 |       import { Transaction } from '@neo-one/smart-contract';
      3 | 
    > 4 |       const for = Transaction['for'];
        |                               ^
      5 |     
"
`;

exports[`Transaction invalid reference - object 1`] = `
"snippetCode.ts (4,15): Builtin properties cannot be referenced

      2 |       import { Transaction } from '@neo-one/smart-contract';
      3 | 
    > 4 |       const { for } = Transaction;
        |               ^
      5 |     
"
`;

exports[`Transaction invalid reference 1`] = `
"snippetCode.ts (4,31): Builtin properties cannot be referenced

      2 |       import { Transaction } from '@neo-one/smart-contract';
      3 | 
    > 4 |       const for = Transaction.for;
        |                               ^
      5 |     
"
`;
